ddl/backfilling_operator.go

```diff
@@ -0,0 +1,511 @@
+	// Build reorg tasks.
+	for _, keyRange := range kvRanges {
+		taskID := taskIDAlloc.alloc()
+		startKey := keyRange.StartKey
+		if len(startKey) == 0 {
+			startKey = prefix
+		}
+		endKey := keyRange.EndKey
+		if len(endKey) == 0 {
+			endKey = prefix.PrefixNext()
+		}
+
+		task := TableScanTask{
+			ID:    taskID,
+			Start: startKey,
+			End:   endKey,
+		}
+		batchTasks = append(batchTasks, task)
+	}
+	return batchTasks
+}
+
+// Close implements Operator interface.
+func (src *TableScanTaskSource) Close() error {
+	return src.errGroup.Wait()
+}
+
+// String implements fmt.Stringer interface.
+func (*TableScanTaskSource) String() string {
+	return "TableScanTaskSource"
+}
+
+// TableScanOperator scans table records in given key ranges from kv store.
+type TableScanOperator struct {
+	*operator.AsyncOperator[TableScanTask, IndexRecordChunk]
+}
+
+// NewTableScanOperator creates a new TableScanOperator.
+func NewTableScanOperator(
+	ctx context.Context,
+	sessPool opSessPool,
+	copCtx *copContext,
+	srcChkPool chan *chunk.Chunk,
+	concurrency int,
+) *TableScanOperator {
+	pool := workerpool.NewWorkerPool(
+		"TableScanOperator",
+		util.DDL,
+		concurrency,
+		func() workerpool.Worker[TableScanTask, IndexRecordChunk] {
+			return &tableScanWorker{
+				ctx:        ctx,
+				copCtx:     copCtx,
+				sessPool:   sessPool,
+				se:         nil,
+				srcChkPool: srcChkPool,
+			}
+		})
+	return &TableScanOperator{
+		AsyncOperator: operator.NewAsyncOperator[TableScanTask, IndexRecordChunk](ctx, pool),
+	}
+}
+
+type tableScanWorker struct {
+	ctx        context.Context
+	copCtx     *copContext
+	sessPool   opSessPool
+	se         *session.Session
+	srcChkPool chan *chunk.Chunk
+}
+
+func (w *tableScanWorker) HandleTask(task TableScanTask, sender func(IndexRecordChunk)) {
+	if w.se == nil {
+		sessCtx, err := w.sessPool.Get()
+		if err != nil {
+			logutil.Logger(w.ctx).Error("copReqSender get session from pool failed", zap.Error(err))
+			sender(IndexRecordChunk{Err: err})
+			return
+		}
+		w.se = session.NewSession(sessCtx)
+	}
+	w.scanRecords(task, sender)
+}
+
+func (w *tableScanWorker) Close() {
+	w.sessPool.Put(w.se.Context)
+}
+
+func (w *tableScanWorker) scanRecords(task TableScanTask, sender func(IndexRecordChunk)) {
+	logutil.Logger(w.ctx).Info("start a table scan task",
+		zap.Int("id", task.ID), zap.String("task", task.String()))
+
+	var idxResult IndexRecordChunk
+	err := wrapInBeginRollback(w.se, func(startTS uint64) error {
+		rs, err := w.copCtx.buildTableScan(w.ctx, startTS, task.Start, task.End)
+		if err != nil {
+			return err
+		}
+		var done bool
+		for !done {
+			srcChk := w.getChunk()
+			done, err = w.copCtx.fetchTableScanResult(w.ctx, rs, srcChk)
+			if err != nil {
+				w.recycleChunk(srcChk)
+				terror.Call(rs.Close)
+				return err
+			}
+			idxResult = IndexRecordChunk{ID: task.ID, Chunk: srcChk, Done: done}
+			sender(idxResult)
+		}
+		return rs.Close()
+	})
+	if err != nil {
+		idxResult.Err = err
```

Comment:
When err, can we stop all operators instead of sending result to the next operator?

Reply:
> When err, can we stop all operators instead of sending result to the next operator?

We can do this later.